### PR TITLE
Bug:54574 - MLCP adjusts concurrency automatically based on DHS Auto-Scaling

### PR DESCRIPTION
**Pull request by yunzvanessa**
1. Add ThreadManager
2. Implement scale-out logic

### CODE REVIEW

File: src/main/java/com/marklogic/contentpump/ThreadManager.java [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+            session = cs.newSession();
+            AdhocQuery query = session.newAdhocQuery(SERVER_MAX_THREADS_QUERY);
+            RequestOptions options = new RequestOptions();
+            options.setDefaultXQueryVersion("1.0-ml");
+            query.setOptions(options);
+            result = session.submitRequest(query);
+
+            if (result.hasNext()) {
+                ResultItem item = result.next();
+                newServerThreads = (int)(maxThreadPercentage *
```
Comment (fsaliba-ml) [inline src/main/java/com/marklogic/contentpump/ThreadManager.java:116]: There's no maxthread cap anymore, the client might not be able to keep up with large clusters? mlcp users should be able to cap newServerThreads.

File: src/main/java/com/marklogic/contentpump/ThreadManager.java [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                        }
+                    } else {
+                        LOG.error(e.getMessage(), e);
+                        job.setJobState(JobStatus.State.FAILED);
+                        return;
+                    }
+                }
+            }
+
+            if (curServerThreads < newServerThreads) {
```
Comment (fsaliba-ml) [inline src/main/java/com/marklogic/contentpump/ThreadManager.java:360]: Just confirming, if an exception is thrown, curServerThreads would be equal to newServerThreads, and no action is taken, right?

**Comment by yunzvanessa** [comment]
Add mlcp scaling in logic and fixed the issues discussed in the last CR.

File: src/main/java/com/marklogic/contentpump/ThreadManager.java [hunk 3]
```diff
@@ -88,9 +99,13 @@ public ThreadManager(LocalJob job, CommandLine cmdline, Command cmd) {
...
         }
+        if (cmdline.hasOption(MAX_THREADS)) {
+            maxThreads = Integer.parseInt(
+                cmdline.getOptionValue(MAX_THREADS));
+        }
         if (cmdline.hasOption(MAX_THREAD_PERCENTAGE)) {
-            maxThreadPercentage = Double.parseDouble(
-                cmdline.getOptionValue(MAX_THREAD_PERCENTAGE));
+            maxThreadPercentage = ((double)Integer.parseInt(
+                cmdline.getOptionValue(MAX_THREAD_PERCENTAGE))) / 100;
```
Comment (fsaliba-ml) [inline src/main/java/com/marklogic/contentpump/ThreadManager.java:108]: Where do we check the value is between 1 and 100?

File: src/main/java/com/marklogic/contentpump/MultithreadedMapper.java [hunk 4]
```diff
@@ -180,6 +183,35 @@ public void createRunners(int numThreads) throws IOException,
         }
     }
 
+    public void stopRunners(int numRunners) {
+        if (numRunners > (threadCount - minRunners)) {
+            LOG.debug("Thread count has reached minimum value: " +
+                minRunners + " for this MultithreadedMapper and the thread " +
```
Comment (fsaliba-ml) [inline src/main/java/com/marklogic/contentpump/MultithreadedMapper.java:189]: Improve debug message, remove MultithreadedMapper

File: src/main/java/com/marklogic/contentpump/ThreadManager.java [hunk 5]
```diff
@@ -229,10 +254,56 @@ public void scaleOutThreadPool() {
                     LOG.error(e.getMessage(), e);
                 }
             } else {
-                return; // vzhang TODO: non Multithreadedmapper case
+                // Safe guard for single-threaded case
+                isScalingDone = true;
```
Comment (fsaliba-ml) [inline src/main/java/com/marklogic/contentpump/ThreadManager.java:258]: A bit confusing, let's fix it

File: src/main/java/com/marklogic/contentpump/ThreadManager.java [hunk 6]
```diff
@@ -229,10 +254,56 @@ public void scaleOutThreadPool() {
...
+                    LOG.debug("Running with MultithreadedMapper. New " +
+                        "thread count for split #" + i + ": " +
+                        newTaskThreads);
+                }
+                task.setThreadCount(newTaskThreads);
+                ((MultithreadedMapper)task.getMapper()).setThreadCount(
+                    newTaskThreads);
+            } else {
+                // Safe guard for single-threaded case
+                isScalingDone = true;
```
Comment (fsaliba-ml) [inline src/main/java/com/marklogic/contentpump/ThreadManager.java:300]: Same fix

File: src/main/java/com/marklogic/contentpump/ThreadManager.java [hunk 7]
```diff
@@ -356,16 +442,15 @@ public void run() {
                     }
                 }
             }
-
             if (curServerThreads < newServerThreads) {
                 // Scale out
+                isScalingDone = false;
```
Comment (fsaliba-ml) [inline src/main/java/com/marklogic/contentpump/ThreadManager.java:447]: We can move  isScalingDone = false; to scaleout and scaclein functions

**Review by fsaliba-ml — COMMENTED** [review]
This doesn't include the fix for the single-threaded case code we discussed (inside the loop), right? Also, what's the max period the thread poller will keep retrying the max thread query, any reason for picking the current numbers? The rest looks great, thanks!

**Comment by yunzvanessa** [comment]
It does include the fix for single-threaded case (bf5f9e1979c084565a4de1ef54a33c7ec700467c). There are two commits in total since you last review. 

For the new retry logic, I just picked a reasonable number. Any suggestions?

**Comment by fsaliba-ml** [comment]
> It does include the fix for single-threaded case ([bf5f9e1](https://github.com/marklogic/marklogic-contentpump/commit/bf5f9e1979c084565a4de1ef54a33c7ec700467c)). There are two commits in total since you last review.
> 
> For the new retry logic, I just picked a reasonable number. Any suggestions?

OK I see the change in the change now ( the else case for if (task.getMapperClass() == (Class)MultithreadedMapper.class) { ) was removed.

**Comment by fsaliba-ml** [comment]
> > It does include the fix for single-threaded case ([bf5f9e1](https://github.com/marklogic/marklogic-contentpump/commit/bf5f9e1979c084565a4de1ef54a33c7ec700467c)). There are two commits in total since you last review.
> > For the new retry logic, I just picked a reasonable number. Any suggestions?
> 
> OK I see the change in the change now ( the else case for if (task.getMapperClass() == (Class)MultithreadedMapper.class) { ) was removed.

For the period, as long as you don't hit the next global polling period.

**Pull request merged**